### ISSUE TITLE
XLocalizer.DB database concurrency problem

### ISSUE THREAD

**Issue by morgrowe** [LazZiya/XLocalizer#28]
Hi Ziya

Hope you're well.

I've been having intermittent issues for the past couple of months with EF complaining about two instances of a DbContext being used at the same time. I recently looked at the logs to try and figure out what's going on and it seems in all cases XLocalizer.DB is involved. The exception I'm referring to is:

```
System.InvalidOperationException: A second operation was started on this context instance before a previous operation completed. This is usually caused by different threads concurrently using the same instance of DbContext.
```

I may be misinterpreting the stack trace, but the exception seems to occur within:

```cs
XLocalizer.DB.EF.EFDbResourceProvider.TryGetValue<TResource>(string, string)
```

every time I see it in my logs.

I've attached a couple of stack traces in case that helps. It's worth noting that this doesn't happen all the time. It's hard to reproduce, but it seems to happen when I excessively refresh a page that uses the `localize-content` tag helper somewhere on it.

If I can provide anymore information or provide additional stack traces, please let me know.

Thanks
Morgan

In application 1:
- XLocalizer.TagHelpers 1.1.0
- XLocalizer 1.0.3
- XLocalizer.DB 1.0.3

[st-1.txt](https://github.com/LazZiya/XLocalizer/files/7609134/st-1.txt)
[st-2.txt](https://github.com/LazZiya/XLocalizer/files/7609135/st-2.txt)
[st-3.txt](https://github.com/LazZiya/XLocalizer/files/7609136/st-3.txt)
[st-4.txt](https://github.com/LazZiya/XLocalizer/files/7609137/st-4.txt)
[st-5.txt](https://github.com/LazZiya/XLocalizer/files/7609138/st-5.txt)
[st-6.txt](https://github.com/LazZiya/XLocalizer/files/7609139/st-6.txt)

In application 2:
- XLocalizer.TagHelpers 1.0.0
- XLocalizer 1.0.0
- XLocalizer.DB 1.0.0

[other-app-st-1.txt](https://github.com/LazZiya/XLocalizer/files/7609190/other-app-st-1.txt)

**Comment by morgrowe** [comment]
I have no evidence, but I have a sneaky feeling it might be something to do with:

https://github.com/LazZiya/XLocalizer.DB/blob/master/XLocalizer.DB/EF/EFDbResourceProvider.cs#L41

```cs
var culture = CultureInfo.CurrentCulture.Name;
```

I normally use `Thread.CurrentThread.CurrentCulture.Name` instead as apparently it's a thread-safe way of accessing the current culture.

**Comment by LazZiya** [comment]
Hi @morgrowe 

I'm fine thanks, hope you are fine too :)

Before I deep dive into the details, do you have below settings as described in XLocalizer.DB docs:
- `ExpressMemeoryCache` enabled in production
- [`DbContext` must be registered as `Transient`](https://docs.ziyad.info/en/XLocalizer/v1.0/setup-db.md#applicationdbcontext-lifetime)
````cs
services.AddDbContext<ApplicationDbContext>(options =>
        options.UseSqlServer(Configuration.GetConnectionString("DefaultConnection")), 
        ServiceLifetime.Transient, 
        ServiceLifetime.Transient);
````


## tldr;
As you already know, the [DbContext is not thread safe](https://docs.microsoft.com/en-us/ef/core/dbcontext-configuration/#:~:text=DbContext%20is%20not%20thread%2Dsafe,to%20use%20the%20context%20instance.), and it is registered as a scoped service.

On the other hand, all localization services `IStringLocalizer`, `IHtmlLocalizer` are singleton services and they expect all their dependencies to be `Singleton` or `Transient`, because a [`Scoped` service can not be consumed inside a `Singleton` service](https://dotnetcoretutorials.com/2018/03/20/cannot-consume-scoped-service-from-singleton-a-lesson-in-asp-net-core-di-scopes/). And here was the challenge to implement  `XLocalizer.DB` life cycle mode!

Since we cannot change the life cycle of localization services, the only solution was to register the `DbContext` as `Transient` because it can be consumed inside `Singleton` localization services.

There are some [drawback of this implementation](https://stackoverflow.com/a/48783504/5519026), but it should not affect `XLocalizer.DB` operations if the `ExpressMemoryCache` is enabled and consider [using multiple DbContexts](https://docs.ziyad.info/en/Miscellaneous/v1.0/multiple-dbcontexts.md) in your app, where the application `DbContext` is registered as `Scoped` by default and `XLocalizer.DB DbContext` can be registered as `Transient` so the other services in your application that are using `DbContext` will not be affected by the `Transient` lifetime of the localization DB.

## Best practices
Here you can find more about [`XLocalizer` best practices](https://docs.ziyad.info/en/XLocalizer/v1.0/best-practice.md).

Best,
Ziya

**Comment by morgrowe** [comment]
Hi Ziya

I'm very good thanks. Glad to hear you're well, too.

Here's the registration of my `DbContext`. Looks like it's setup to be Transient as your docs advise:

```cs
var connectionString = "MyConnectionString";
var migrationsAssembly = typeof(ManagerDbContext).Assembly.FullName;

services.AddDbContext<ManagerDbContext>(options =>
{
    options.UseSqlServer(connectionString, b => b.MigrationsAssembly(migrationsAssembly));
}
, ServiceLifetime.Transient
ServiceLifetime.Transient);
```

And here are my `AddXDbLocalizer` options:

```cs
.AddXDbLocalizer<ManagerDbContext, DummyTranslator, AppLocalizationResource>((opts) =>
{
    opts.AutoTranslate = false;
    opts.LocalizeDefaultCulture = true;
});
```

I think I was using `ExpressMemeoryCache`, but I think I disabled it before `LocalizeDefaultCulture` was added. I may need to add that back for some perf gains. :)

Thanks for the link to the pros and cons of using a Transient `DbContext`. I noticed that I wasn't able to return tracked entities from one class to another, so that explains why!

Your suggestion of having two DbContexts is a welcome one. Despite having multiple `DbContext`s in this application, it didn't even occur to me to have one just for XDbLocalization. Should I try implementing that first to see if that resolves the issue I posted about? Or would you like to do some investigation first with the information I've provided?

Thanks
Morgan

**Comment by LazZiya** [comment]
So `ExpressMemoryCache` will help a lot in this case. Additionally, I would always recommend a dedicated DB for localization services, this will keep your other application services not affected by the `Transient` implementation of localization DB service, and you will have a clean DB architecture.

For the issue, it will take some time, I can't promise to bring a solution in a short time, but the above recommendations would help to avoid the problem :)

**Comment by morgrowe** [comment]
Ok, I will enable `ExpressMemoryCache` and decouple the XDbLocalizer tables from my application's DbContext in order to load just the XDbLocalizer's DbContext using Transient scoping. That'll let me add my application's DbContext back as Scoped.

Many thanks for agreeing to launch an investigation. In the meantime, I'll share any exceptions relating to this issue with you should they pop up again over the next few days. :)

Thank you very much
Morgam

**Comment by LazZiya** [comment]
Thank you too for your cooperation :)

Best,
Ziya

**Comment by morgrowe** [comment]
Hi Ziya

I didn't have much time to do development this week, but I did get two exceptions relating to this issue yesterday afternoon. They happened a couple of milliseconds after each other like the other ones. These occurred after setting `UseExpressMemoryCache = true` and creating a separate `DbContext` for the resources (`LocalizationDbContext`).

[caforb-ex-wk2.txt](https://github.com/LazZiya/XLocalizer/files/7650207/caforb-ex-wk2.txt)

Cheers
Morgan

**Comment by morgrowe** [comment]
Hi Ziya

Hope you're well. I can't believe almost a year has passed since I brought this up. 

We're still getting the exception (even in Release mode running on IIS). Did you have a chance to look into this?

Thanks
Morgan

**Comment by LazZiya** [comment]
Hi Morgan

I am fine thanks, hope you also fine.

Yea time is passing fast! also projects are growing rapidly...
Unfortunately I couldn't really afford time to deep dive into this issue :(

Any PR are welcome :)

Best regards,
Ziya

**Comment by morgrowe** [comment]
Hi Ziya

Ok, no worries. I had a look into this a couple of months ago, but I couldn't find a solution.

I will have another look. If I fix it, I'll do a pull request.

Thanks
Morgan

**Comment by Sayed-Mahmoud** [comment]
I have the same problem when using Arabic!
.Net Core 5 Razor Pages with EF Core

**Comment by ScubaAddict1** [comment]
HI @morgrowe , did you manage to find a solution? Im running .net9 and noticed some strange 503 error, mainly when crawlers like google crawl the page, so suspect might be a similar issue. So thinking of converting from db to xml. Just need to find a way to export db to xml instead of resx. :)

**Comment by LazZiya** [comment]
Hi all,

Long time didn't interact with this project. Finally I could afford some time today to deep dive into this issue.

I released a preview version that should resolve the concurrency issue by using `IDbContextFactory` pattern instead of directly using the `DbContext`.

You can read the full steps in the docs: [Fix for concurrency issue](https://docs.ziyad.info/en/XLocalizer/v1.0/setup-db-factory-pattern.md)

Kindly let me know if it helps, so I can release the nugets as a stable version :)

**Comment by morgrowe** [comment]
Hi @ScubaAddict1. Yeah, I was getting the same errors--I could never track down specifically what was causing them, so we switched from the DB provider to the XML one. It has worked perfectly with the XML provider. 

However, with @LazZiya's latest update, that may no longer be necessary. Thanks @LazZiya :)

**Comment by LazZiya** [comment]
Closing this issue as the last update should fix it.

Feel free to reopen if you still have problems.

Best regards,
Ziya
